File: frontend/mobile/lib/core/services/achievements_service.dart
import 'package:dio/dio.dart';
import '../../features/achievements/data/models/achievement_model.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';

/// Service for managing achievements, badges, levels, and challenges
class AchievementsService {
  final DioClient _dioClient;

  AchievementsService(this._dioClient);

  /// Get achievements dashboard with all data
  Future<AchievementsDashboard> getAchievementsDashboard() async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.achievementsDashboard,
      );
      return AchievementsDashboard.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get all achievements (locked and unlocked)
  Future<List<Achievement>> getAllAchievements({
    AchievementCategory? category,
    AchievementTier? tier,
    bool? isUnlocked,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (category != null) {
        queryParams['category'] = category.toString().split('.').last;
      }
      if (tier != null) {
        queryParams['tier'] = tier.toString().split('.').last;
      }
      if (isUnlocked != null) {
        queryParams['is_unlocked'] = isUnlocked;
      }

      final response = await _dioClient.get(
        ApiEndpoints.achievements,
        queryParameters: queryParams,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => Achievement.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get specific achievement details
  Future<Achievement> getAchievement(String achievementId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.achievements}/$achievementId',
      );
      return Achievement.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get user level and points information
  Future<UserLevel> getUserLevel() async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.userLevel,
      );
      return UserLevel.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get active challenges (daily and weekly)
  Future<List<Challenge>> getActiveChallenges({
    ChallengeType? type,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (type != null) {
        queryParams['type'] = type.toString().split('.').last;
      }

      final response = await _dioClient.get(
        ApiEndpoints.challenges,
        queryParameters: queryParams,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => Challenge.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get specific challenge details
  Future<Challenge> getChallenge(String challengeId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.challenges}/$challengeId',
      );
      return Challenge.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Update challenge progress
  Future<Challenge> updateChallengeProgress(
    String challengeId,
    int progressValue,
  ) async {
    try {
      final response = await _dioClient.post(
        '${ApiEndpoints.challenges}/$challengeId/progress',
        data: {
          'progress_value': progressValue,
        },
      );
      return Challenge.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get achievement statistics
  Future<AchievementStats> getAchievementStats() async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.achievementStats,
      );
      return AchievementStats.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get motivational reminders
  Future<List<MotivationalReminder>> getReminders({
    bool? isActive,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (isActive != null) {
        queryParams['is_active'] = isActive;
      }

      final response = await _dioClient.get(
        ApiEndpoints.achievementReminders,
        queryParameters: queryParams,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => MotivationalReminder.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Create or update a reminder
  Future<MotivationalReminder> saveReminder(MotivationalReminder reminder) async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.achievementReminders,
        data: reminder.toJson(),
      );
      return MotivationalReminder.fromJson(response.data);
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Delete a reminder
  Future<void> deleteReminder(String reminderId) async {
    try {
      await _dioClient.delete(
        '${ApiEndpoints.achievementReminders}/$reminderId',
      );
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Share achievement on social media
  Future<Map<String, dynamic>> shareAchievement(
    ShareAchievementRequest request,
  ) async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.shareAchievement,
        data: request.toJson(),
      );
      return response.data as Map<String, dynamic>;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get achievement unlock history
  Future<List<AchievementUnlockNotification>> getUnlockHistory({
    int? limit,
    DateTime? since,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (limit != null) {
        queryParams['limit'] = limit;
      }
      if (since != null) {
        queryParams['since'] = since.toIso8601String();
      }

      final response = await _dioClient.get(
        ApiEndpoints.achievementUnlockHistory,
        queryParameters: queryParams,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => AchievementUnlockNotification.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Manually trigger achievement check (for testing/debugging)
  Future<List<Achievement>> checkAchievements() async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.checkAchievements,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => Achievement.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  /// Get leaderboard (if social features are enabled)
  Future<List<Map<String, dynamic>>> getLeaderboard({
    String? timeframe, // 'daily', 'weekly', 'monthly', 'all_time'
    int? limit,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (timeframe != null) {
        queryParams['timeframe'] = timeframe;
      }
      if (limit != null) {
        queryParams['limit'] = limit;
      }

      final response = await _dioClient.get(
        ApiEndpoints.achievementLeaderboard,
        queryParameters: queryParams,
      );
      
      return (response.data as List<dynamic>)
          .map((json) => json as Map<String, dynamic>)
          .toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  Exception _handleError(DioException error) {
    if (error.response != null) {
      final statusCode = error.response!.statusCode;
      final message = error.response!.data['message'] ?? 'Unknown error occurred';
      
      switch (statusCode) {
        case 400:
          return Exception('Bad request: $message');
        case 401:
          return Exception('Unauthorized: Please login again');
        case 404:
          return Exception('Not found: $message');
        case 500:
          return Exception('Server error: $message');
        default:
          return Exception('Error: $message');
      }
    } else {
      return Exception('Network error: ${error.message}');
    }
  }
}

File: frontend/nextjs-app/src/lib/services/achievements-service.ts
/**
 * Achievements service for Next.js web app
 * Handles all achievements, badges, levels, and challenges API calls
 */

import axiosClient from '../api/axios-client';
import {
  Achievement,
  AchievementCategory,
  AchievementTier,
  AchievementsDashboard,
  AchievementStats,
  AchievementUnlockNotification,
  Challenge,
  ChallengeType,
  LeaderboardEntry,
  MotivationalReminder,
  ShareAchievementRequest,
  UserLevel,
} from '@/types/achievements';

const ACHIEVEMENTS_BASE = '/api/achievements';

export class AchievementsService {
  /**
   * Get achievements dashboard with all data
   */
  static async getAchievementsDashboard(): Promise<AchievementsDashboard> {
    const response = await axiosClient.get<AchievementsDashboard>(
      `${ACHIEVEMENTS_BASE}/dashboard`
    );
    return response.data;
  }

  /**
   * Get all achievements (locked and unlocked)
   */
  static async getAllAchievements(params?: {
    category?: AchievementCategory;
    tier?: AchievementTier;
    isUnlocked?: boolean;
  }): Promise<Achievement[]> {
    const response = await axiosClient.get<Achievement[]>(
      `${ACHIEVEMENTS_BASE}/achievements`,
      { params }
    );
    return response.data;
  }

  /**
   * Get specific achievement details
   */
  static async getAchievement(achievementId: string): Promise<Achievement> {
    const response = await axiosClient.get<Achievement>(
      `${ACHIEVEMENTS_BASE}/achievements/${achievementId}`
    );
    return response.data;
  }

  /**
   * Get user level and points information
   */
  static async getUserLevel(): Promise<UserLevel> {
    const response = await axiosClient.get<UserLevel>(
      `${ACHIEVEMENTS_BASE}/level`
    );
    return response.data;
  }

  /**
   * Get active challenges (daily and weekly)
   */
  static async getActiveChallenges(params?: {
    type?: ChallengeType;
  }): Promise<Challenge[]> {
    const response = await axiosClient.get<Challenge[]>(
      `${ACHIEVEMENTS_BASE}/challenges`,
      { params }
    );
    return response.data;
  }

  /**
   * Get specific challenge details
   */
  static async getChallenge(challengeId: string): Promise<Challenge> {
    const response = await axiosClient.get<Challenge>(
      `${ACHIEVEMENTS_BASE}/challenges/${challengeId}`
    );
    return response.data;
  }

  /**
   * Update challenge progress
   */
  static async updateChallengeProgress(
    challengeId: string,
    progressValue: number
  ): Promise<Challenge> {
    const response = await axiosClient.post<Challenge>(
      `${ACHIEVEMENTS_BASE}/challenges/${challengeId}/progress`,
      { progress_value: progressValue }
    );
    return response.data;
  }

  /**
   * Get achievement statistics
   */
  static async getAchievementStats(): Promise<AchievementStats> {
    const response = await axiosClient.get<AchievementStats>(
      `${ACHIEVEMENTS_BASE}/stats`
    );
    return response.data;
  }

  /**
   * Get motivational reminders
   */
  static async getReminders(params?: {
    isActive?: boolean;
  }): Promise<MotivationalReminder[]> {
    const response = await axiosClient.get<MotivationalReminder[]>(
      `${ACHIEVEMENTS_BASE}/reminders`,
      { params }
    );
    return response.data;
  }

  /**
   * Create or update a reminder
   */
  static async saveReminder(
    reminder: MotivationalReminder
  ): Promise<MotivationalReminder> {
    const response = await axiosClient.post<MotivationalReminder>(
      `${ACHIEVEMENTS_BASE}/reminders`,
      reminder
    );
    return response.data;
  }

  /**
   * Delete a reminder
   */
  static async deleteReminder(reminderId: string): Promise<void> {
    await axiosClient.delete(`${ACHIEVEMENTS_BASE}/reminders/${reminderId}`);
  }

  /**
   * Share achievement on social media
   */
  static async shareAchievement(
    request: ShareAchievementRequest
  ): Promise<{ shareUrl?: string; message: string }> {
    const response = await axiosClient.post<{
      shareUrl?: string;
      message: string;
    }>(`${ACHIEVEMENTS_BASE}/share`, request);
    return response.data;
  }

  /**
   * Get achievement unlock history
   */
  static async getUnlockHistory(params?: {
    limit?: number;
    since?: string;
  }): Promise<AchievementUnlockNotification[]> {
    const response = await axiosClient.get<AchievementUnlockNotification[]>(
      `${ACHIEVEMENTS_BASE}/unlock-history`,
      { params }
    );
    return response.data;
  }

  /**
   * Manually trigger achievement check (for testing/debugging)
   */
  static async checkAchievements(): Promise<Achievement[]> {
    const response = await axiosClient.post<Achievement[]>(
      `${ACHIEVEMENTS_BASE}/check`
    );
    return response.data;
  }

  /**
   * Get leaderboard (if social features are enabled)
   */
  static async getLeaderboard(params?: {
    timeframe?: 'daily' | 'weekly' | 'monthly' | 'all_time';
    limit?: number;
  }): Promise<LeaderboardEntry[]> {
    const response = await axiosClient.get<LeaderboardEntry[]>(
      `${ACHIEVEMENTS_BASE}/leaderboard`,
      { params }
    );
    return response.data;
  }
}

export default AchievementsService;
